src/braket/default_simulator/openqasm/interpreter.py

```diff
@@ -0,0 +1,595 @@
+    BitType,
+    BooleanLiteral,
+    BranchingStatement,
+    Cast,
+    ClassicalArgument,
+    ClassicalAssignment,
+    ClassicalDeclaration,
+    ConstantDeclaration,
+    DiscreteSet,
+    FloatLiteral,
+    ForInLoop,
+    FunctionCall,
+    GateModifierName,
+    Identifier,
+    Include,
+    IndexedIdentifier,
+    IndexExpression,
+    IntegerLiteral,
+    IODeclaration,
+    IOKeyword,
+    Pragma,
+    Program,
+    QASMNode,
+    QuantumGate,
+    QuantumGateDefinition,
+    QuantumGateModifier,
+    QuantumMeasurement,
+    QuantumPhase,
+    QuantumReset,
+    QuantumStatement,
+    QubitDeclaration,
+    RangeDefinition,
+    ReturnStatement,
+    SizeOf,
+    SubroutineDefinition,
+    UnaryExpression,
+    WhileLoop,
+)
+
+from braket.default_simulator.openqasm.data_manipulation import (
+    builtin_constants,
+    builtin_functions,
+    cast_to,
+    convert_range_def_to_range,
+    convert_to_gate,
+    create_empty_array,
+    evaluate_binary_expression,
+    evaluate_unary_expression,
+    get_ctrl_modifiers,
+    get_elements,
+    get_identifier_name,
+    get_operator_of_assignment_operator,
+    index_expression_to_indexed_identifier,
+    invert,
+    is_controlled,
+    is_inverted,
+    is_literal,
+    modify_body,
+    singledispatchmethod,
+    wrap_value_into_literal,
+)
+from braket.default_simulator.openqasm.program_context import ProgramContext
+
+
+class Interpreter:
+    """
+    The interpreter is responsible for visiting the AST of an OpenQASM program, as created
+    by the parser, and building a braket.default_simulator.openqasm.circuit.Circuit to hand
+    off to a simulator e.g. braket.default_simulator.state_vector_simulator.StateVectorSimulator.
+
+    The interpreter keeps track of all state using a ProgramContext object. The main entry point
+    is build_circuit(), which returns the built circuit. An alternative entry poitn, run() returns
+    the ProgramContext object, which can be used for debugging or other customizability.
+    """
+
+    def __init__(self, context: Optional[ProgramContext] = None, logger: Optional[Logger] = None):
+        # context keeps track of all state
+        self.context = context or ProgramContext()
+        self.logger = logger or getLogger(__name__)
+
+    def build_circuit(
+        self, source: str, inputs: Optional[Dict[str, io_type]] = None, is_file: bool = False
+    ):
+        """Interpret an OpenQASM program and build a Circuit IR."""
+        if inputs:
+            self.context.load_inputs(inputs)
+
+        if is_file:
+            with open(source, "r") as f:
+                source = f.read()
+
+        program = parse(source)
+        self.visit(program)
+        return self.context.circuit
+
+    def run(self, source, inputs=None, is_file=False):
+        """Interpret an OpenQASM program and return the program state"""
+        self.build_circuit(source, inputs, is_file)
+        return self.context
+
+    @singledispatchmethod
```

Comment:
How are we planning to handle the fact that this doesn't work with Python 3.7?

Reply:
Responded above, but I added an implementation [here](https://github.com/aws/amazon-braket-default-simulator-python/blob/openqasm-local-simulator/src/braket/default_simulator/openqasm/data_manipulation.py#L640)